### PR TITLE
[RFC-3086] Consider out-of-bound depths of `count`

### LINKED ISSUE #111905: `${count()}` disregards depth parameter when captured repetition is empty
I tried this code:

```rust
#![feature(macro_metavar_expr)]

macro_rules! foo {
    ($($t:ident)*) => { ${count(t, 4294967296)} }; 
}

macro_rules! bar {
    ( $( { $( [ $( ( $( $t:ident )* ) )* ] )* } )* ) => { ${count(t, 4294967296)} }
}

fn test() {
    foo!();
    bar!( { [] [] } );
}
```

I expected to see this happen: Fails to expand both macro calls, because the depth parameter (4294967296) is out of bounds for `$t` fragment in both macros.

Instead, this happened: Compiled successfully, both macro calls expand to `0`.

### Meta

`rustc --version --verbose`:
```
1.71.0-nightly (2023-05-23 5ea3f0ae08c07472239a)
```

cc #83527
@rustbot label requires-nightly F-macro-metavar-expr

### PR DESCRIPTION
**Pull request by c410-f3r**
Fix #111905

In the matching of macro calls and their respective declarations (transcribe), a warning is issued if `count` has a depth greater than the number of nested `NamedMatch`s through be verification of `MatchedSeq` with empty elements.

Doesn't affect `( $( { $( [ $( ( $( $t:ident )* ) )* ] )* } )* ) => { ${count(t, 1)} }` called with `bar!( { [] [] } )` which will still continue to output `2`.

### CODE REVIEW

**Comment by c410-f3r** [comment]
r? compiler

**Comment by oli-obk** [comment]
I'm on vacation and won't get to this before the 19th

Also cc @petrochenkov for macro stuff

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 1]
```diff
@@ -421,22 +421,22 @@ fn lockstep_iter_size(
 /// * `[ $( ${count(foo, 0)} ),* ]` will be the same as `[ $( ${count(foo)} ),* ]`
 /// * `[ $( ${count(foo, 1)} ),* ]` will return an error because `${count(foo, 1)}` is
 ///   declared inside a single repetition and the index `1` implies two nested repetitions.
-fn count_repetitions<'a>(
```
Comment (oli-obk) [inline compiler/rustc_expand/src/mbe/transcribe.rs:424]: why did you rename the lifetime? It's not clearly better to me

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 2]
```diff
@@ -448,6 +448,12 @@ fn count_repetitions<'a>(
                 }
             }
             MatchedSeq(named_matches) => {
+                if named_matches.is_empty() {
+                    return Err(cx.struct_span_err(
+                        sp.entire(),
+                        "related fragment that refers the `count` meta-variable expression was not found",
+                    ));
```
Comment (oli-obk) [inline compiler/rustc_expand/src/mbe/transcribe.rs:455]: You could add a structured diagnostic for this new error

**Comment by oli-obk** [comment]
@craterbot check

**Comment by c410-f3r** [comment]
@oli-obk `count` is part of the nightly `macro_metavar_expr` feature and this PR is fixing a logical error that should be rare to appear in practice.

> needs-fcp: This change is insta-stable, so needs a completed FCP to proceed.

That said, I think a FCP isn't necessary.

**Comment by oli-obk** [comment]
Oh I didn't realize that. I just saw the behaviour change...

@craterbot cancel

**Comment by oli-obk** [comment]
@craterbot abort

File: compiler/rustc_expand/src/mbe/transcribe.rs (hunk 1 above)
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs:424]: Hehehehe, it was a "good intention". `'a` doesn't have a semantic value but neither does `'any`. 

Reverted

File: compiler/rustc_expand/src/mbe/transcribe.rs (hunk 2 above)
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs]: Done

File: tests/ui/macros/rfc-3086-metavar-expr/issue-111905.stderr [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  --> $DIR/issue-111905.rs:4:26
+   |
+LL |     ($($t:ident)*) => { ${count(t, 4294967296)} };
+   |                          ^^^^^^^^^^^^^^^^^^^^^^
+
+error: related fragment that refers the `count` meta-variable expression was not found
+  --> $DIR/issue-111905.rs:9:60
+   |
+LL |     ( $( { $( [ $( ( $( $t:ident )* ) )* ] )* } )* ) => { ${count(t, 4294967296)} }
+   |                                                            ^^^^^^^^^^^^^^^^^^^^^^
```
Comment (oli-obk) [inline tests/ui/macros/rfc-3086-metavar-expr/issue-111905.stderr:11]: Do we have enough information to point to the `$` in the pattern which corresponds to the current level?

**Comment by JohnCSimon** [comment]
@c410-f3r 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

File: tests/ui/macros/rfc-3086-metavar-expr/issue-111905.stderr (hunk 3 above)
Comment (c410-f3r) [reply tests/ui/macros/rfc-3086-metavar-expr/issue-111905.stderr:11]: Well, I wasn't able to capture the pattern span after ~1 hour of attempts

**Comment by c410-f3r** [comment]
Sorry for the ~2 months delay 🫠

I couldn't reproduce the CI error locally with `--build=i686 ...` and it is odd to see a UI test failing in an i686 environment

**Comment by c410-f3r** [comment]
Well, the matter needs more investigation on my side

**Comment by ehuss** [comment]
@bors r-

bors synchronize queue fix

**Comment by Dylan-DPC** [comment]
@c410-f3r any updates on this?

**Comment by c410-f3r** [comment]
Thank you @oli-obk 

I spent several hours trying to find a solution and nothing worked. Now I don't have the time or motivation to continue perusing this PR.
